src/measure_theory/borel_hierarchy.lean

```diff
@@ -0,0 +1,505 @@
+/-
+Copyright (c) 2022 Pedro Sánchez Terraf. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pedro Sánchez Terraf
+-/
+import set_theory.cardinal.continuum
+import measure_theory.measurable_space_def
+import set_theory.cardinal.cofinality
+import tactic.induction
+
+/-!
+# The Borel hierarchy
+
+In this file we define recursively define the classical Borel hierarchy of sets.
+It is developed as a purely set-theoretic definition here, and it is used to reprove
+the result on the cardinal bound for σ-algebras, previously formalized by S. Gouëzel.
+
+## Main definitions
+
+- `sigma0_pi0_rec`: Recursive definition of the hierarchy, not intended for direct use.
+- `sigma0`, `pi0` : The classical pointclasses, obtained from `sigma0_pi0_rec`
+using its `bool` argument.
+
+## Implementation notes
+
+Traditionally, pointclasses $\Sigma^0_\alpha$ and $\Pi^0_\alpha$ for positive
+values of the ordinal $\alpha$. Here the definition is extended in such a way that
+$\Sigma^0_0 = \emptyset$ (which results in more generality for some lemmas) and
+$\Pi^0_0$ coincides with the set of generators.
+
+The `inductive` definition of `sigma0_pi0_rec` was suggested by Junyan Xu.
+
+## References
+
+The definition of the hierarchy was extracted from [kechris1995].
+-/
+
+universe u
+
+namespace pointclasses
+
+section sigma0_pi0
+
+/-!
+### Σ and Π pointclasses
+
+This section includes basic definitions and API.
+-/
+open set
+
+variables {α : Type u} (s : set (set α)) (i k : ordinal.{u})
+
+/--
+Simultaneous recursive definition of Σ⁰ᵢ and Π⁰ᵢ pointclasses by recursion
+on ordinals (a variant of 11.B in Kechris, [Classical Descriptive Set Theory][kechris1995]).
+
+The main difference is that the hierarchy starts at level 0: Π⁰₀ are intended to
+be basic open sets (augmented with `∅` and `univ`) and Σ⁰₀ is the empty family.
+-/
+inductive sigma0_pi0_rec {α : Type u} (s : set (set α)) :
+  ordinal.{u} → bool → set α → Prop
+| basic (x ∈ s) : sigma0_pi0_rec 0 tt x
+| empty         : sigma0_pi0_rec 0 tt ∅
+| univ          : sigma0_pi0_rec 0 tt univ
+| compl {i x}   : sigma0_pi0_rec i ff x → sigma0_pi0_rec i tt xᶜ
+| union {i} (f : ℕ → set α) (g : ℕ → ordinal.{u}) :
+    (∀ n, g n < i) → (∀ n, sigma0_pi0_rec (g n) tt (f n)) → sigma0_pi0_rec i ff (⋃ n, f n)
+
+/--
+The family of (boldface) Σ⁰ᵢ pointsets, which are countable unions of Π⁰ⱼ sets
```

Comment:
The definition depends on two parameters `s` and `i`, but  the docstring only mentions `i`. And I am not sure it is understandable in a self-contained way... Same thing below.